group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -3981,6 +4105,30 @@ public CoordinatorResult<ShareGroupHeartbeatResponseData, CoordinatorRecord> sha
             request.subscribedTopicNames());
     }
 
+    /**
+     * Handles an initialize share group state request. This is usually part of
+     * shareGroupHeartbeat code flow.
+     * @param groupId The group id corresponding to the share group whose share partitions have been initialized.
+     * @param topicPartitionMap Map representing topic partition data to be added to the share state partition metadata.
+     *
+     * @return A Result containing ShareGroupStatePartitionMetadata records and Void response.
+     */
+    public CoordinatorResult<Void, CoordinatorRecord> initializeShareGroupState(
+        String groupId,
+        Map<Uuid, Map.Entry<String, List<Integer>>> topicPartitionMap
+    ) {
+        // Should be present
+        ShareGroup group = (ShareGroup) groups.get(groupId);
+        if (topicPartitionMap == null || topicPartitionMap.isEmpty()) {
+            return new CoordinatorResult<>(List.of(), null);
+        }
+
+        return new CoordinatorResult<>(
+            List.of(newShareGroupPartitionMetadataRecord(group.groupId(), topicPartitionMap, Map.of())),
```

Comment:
This is incorrect in the case where the number of partitions has increased. Let's say that the T1 has increased from 1 to 4 partitions. The `topicPartitionMap` will contain `{topicId, {"T1", [1,2,3]}}`. However, the `ShareGroupPartitionMetadataRecord` should contain the complete list of partitions, not just the new ones. Because the `__consumer_offsets` topic is compacted and these records are keyed by group ID, only one such record can be expected to remain present. As a result, the most recent record needs to contain the most up-to-date list of all initialized topic-partitions.